pkg/flow/internal/controller/loader.go

```diff
@@ -71,45 +71,21 @@ func NewLoader(globals ComponentGlobals) *Loader {
 // The provided parentContext can be used to provide global variables and
 // functions to components. A child context will be constructed from the parent
 // to expose values of other components.
-func (l *Loader) Apply(parentScope *vm.Scope, blocks []*ast.BlockStmt, configBlocks []*ast.BlockStmt) diag.Diagnostics {
+func (l *Loader) Apply(parentScope *vm.Scope, componentNodeBlocks []*ast.BlockStmt, configBlocks []*ast.BlockStmt) diag.Diagnostics {
 	start := time.Now()
 	l.mut.Lock()
 	defer l.mut.Unlock()
 	l.cm.controllerEvaluation.Set(1)
 	defer l.cm.controllerEvaluation.Set(0)
 
-	var (
-		diags    diag.Diagnostics
-		newGraph dag.Graph
-	)
-
-	// Pre-populate graph with a ConfigNode.
-	c, configBlockDiags := NewConfigNode(configBlocks, l.globals, l.isModule())
-	diags = append(diags, configBlockDiags...)
-	newGraph.Add(c)
-
-	// Handle the rest of the graph as ComponentNodes.
-	populateDiags := l.populateGraph(&newGraph, blocks)
-	diags = append(diags, populateDiags...)
-
-	wireDiags := l.wireGraphEdges(parentScope, &newGraph)
-	diags = append(diags, wireDiags...)
-
-	// Validate graph to detect cycles
-	err := dag.Validate(&newGraph)
-	if err != nil {
-		diags = append(diags, multierrToDiags(err)...)
+	newGraph, diags := l.loadNewGraph(parentScope, componentNodeBlocks, configBlocks)
+	if diags.HasErrors() {
```

Comment:
Hmm, yeah, I'm struggling to tell if this matters. Ideally we collect as many diagnostics as possible so people don't have to keep reloading a bunch of times to see another set of errors, but it also might be a good idea to do an initial return if building the DAG or validating it had any issues, since it might compound unrelated errors when the graph is being walked.

Maybe this is fine as-is for now, and we can revisit if we get negative feedback?